src/main/webapp/app/shared/util/firebase/firebase-review-utils.tsx

```diff
@@ -61,52 +65,47 @@ export class BaseReviewLevel {
   }
 }
 
+export type MetaReviewLevelParams = Omit<BaseReviewLevelParams, 'reviewLevelType'>;
+
 export class MetaReviewLevel extends BaseReviewLevel {
-  constructor(title: string, currentValPath: string) {
-    super(title, currentValPath);
-    this.reviewLevelType = ReviewLevelType.META;
+  constructor(params: MetaReviewLevelParams) {
+    super({ reviewLevelType: ReviewLevelType.META, ...params });
   }
 }
 
-export class ReviewLevel extends BaseReviewLevel {
-  currentVal: any;
+export type ReviewInfo = {
   reviewPath: string;
   review: Review;
   lastReviewedString: string;
-  reviewAction: ReviewAction;
   uuid: string;
-  newState?: HistoryRecordState;
+  reviewAction?: ReviewAction;
+  diffMethod?: DiffMethod;
+};
+
+export type HistoryData = {
   oldState?: HistoryRecordState;
-  deleteIndex?: number;
-  diffMethod?: DiffMethod = DiffMethod.CHARS;
-
-  constructor(
-    title: string,
-    currentValPath: string,
-    currentVal: any,
-    reviewPath: string,
-    review: Review,
-    lastReviewedString: string,
-    uuid: string,
-    newState: HistoryRecordState,
-    oldState: HistoryRecordState,
-    isUnderCreationOrDeletion = false
-  ) {
-    super(title, currentValPath, isUnderCreationOrDeletion);
+  newState?: HistoryRecordState;
+};
+
+export type ReviewLevelParams = Omit<BaseReviewLevelParams, 'reviewLevelType'> & {
+  currentVal: any;
+  reviewInfo: ReviewInfo;
+  historyData: HistoryData;
+};
+
+export class ReviewLevel extends BaseReviewLevel {
+  currentVal: any;
+  reviewInfo: ReviewInfo;
+  historyData: HistoryData;
+
+  constructor({ currentVal, reviewInfo, historyData, ...baseReviewLevelParams }: ReviewLevelParams) {
+    super({ ...baseReviewLevelParams, reviewLevelType: ReviewLevelType.REVIEWABLE });
     this.currentVal = currentVal;
-    this.reviewPath = reviewPath;
-    this.review = review;
-    this.lastReviewedString = lastReviewedString;
-    this.uuid = uuid;
-    this.reviewLevelType = ReviewLevelType.REVIEWABLE;
-    this.reviewAction = getReviewAction(review, reviewPath);
-    if (newState !== undefined) {
-      this.newState = newState;
-    }
-    if (oldState !== undefined) {
-      this.oldState = oldState;
+    if (!reviewInfo.reviewAction) {
```

Comment:
Wondering why need this check. I haven't seen a use case where `reviewAction` is set

Reply:
Yeah, this can be removed